Optimizer assertion error on Cats Effect tests in 1.20.2 (Scala 2.12)

This one has been blocking CE upgrades for a while. I'm bypassing it for now by turning off optimization in the tests but obviously that's not ideal.

<details>

```
[error] (testsJS / Test / fastLinkJS) org.scalajs.linker.frontend.optimizer.OptimizerCore$OptimizeException: The Scala.js optimizer crashed while optimizing constructor cats.effect.unsafe.SchedulerSuite.<init>;V: java.lang.AssertionError: assertion failed: Found Assign in expression position
[error] Methods attempted to inline:
[error] * constructor munit.FunSuite.<init>;V
[error] * munit.internal.MacroCompat$CompileErrorMacro.$init$;V
[error] * munit.FunFixtures.$init$;V
[error] * munit.TestOptionsConversions.$init$;V
[error] * munit.TestTransforms.$init$;V
[error] * munit.SuiteTransforms.$init$;V
[error] * munit.ValueTransforms.$init$;V
[error] * munit.BaseFunSuite.$init$;V
[error] * munit.FunSuite.munit$BaseFunSuite$_setter_$munitTestsBuffer_$eq;Lscala.collection.mutable.ListBuffer;V
[error] * cats.effect.kernel.testkit.ParallelFGenerators.$init$;V
[error] * cats.effect.kernel.testkit.OutcomeGenerators.$init$;V
[error] * cats.effect.RunnersPlatform.$init$;V
[error] * cats.effect.DetectPlatform.$init$;V
[error] * cats.effect.Runners.$init$;V
[error] * cats.effect.BaseSuite.$init$;V
[error] * munit.FunSuite.testOptionsFromString;Ljava.lang.String;Lmunit.Location;Lmunit.TestOptions
[error] * scala.Function0.$init$;V
[error] * cats.effect.unsafe.SchedulerSuite.real;Lmunit.TestOptions;Lscala.Function0;Lmunit.Location;V
[error] * cats.effect.Runners.real;Lmunit.TestOptions;Lscala.Function0;Lmunit.Location;V
[error] * munit.FunSuite.test;Lmunit.TestOptions;Lscala.Function0;Lmunit.Location;V
[error] * munit.BaseFunSuite.test;Lmunit.TestOptions;Lscala.Function0;Lmunit.Location;V
[error] * munit.FunSuite.munitTestsBuffer;Lscala.collection.mutable.ListBuffer
[error] * munit.TestOptions.name;Ljava.lang.String
[error] * munit.TestOptions.tags;Lscala.collection.immutable.Set
[error] * private munit.BaseFunSuite.waitForCompletion;Lscala.Function0;Lscala.concurrent.Future
[error] * cats.effect.unsafe.SchedulerSuite.munitTimeout;Lscala.concurrent.duration.Duration
[error] * cats.effect.Runners.munitTimeout;Lscala.concurrent.duration.Duration
[error] * cats.effect.unsafe.SchedulerSuite.executionTimeout;Lscala.concurrent.duration.FiniteDuration
[error] * munit.Suite.munitExecutionContext;Lscala.concurrent.ExecutionContext
[error] * private munit.Suite.parasiticExecutionContext;Lscala.concurrent.ExecutionContext
[error] * munit.internal.PlatformCompat$.waitAtMost;Lscala.Function0;Lscala.concurrent.duration.Duration;Lscala.concurrent.ExecutionContext;Lscala.concurrent.Future
[error] * scala.concurrent.Promise$.apply;Lscala.concurrent.Promise
[error] * scala.concurrent.duration.FiniteDuration.isFinite;Z
[error] * scala.concurrent.duration.FiniteDuration.toMillis;J
[error] * scala.concurrent.duration.FiniteDuration.unit;Ljava.util.concurrent.TimeUnit
[error] * scala.concurrent.duration.FiniteDuration.length;J
[error] * static org.scalajs.linker.runtime.RuntimeLong.toDouble;I;I;D
[error] * org.scalajs.linker.runtime.RuntimeLong$.toDouble;I;I;D
[error] * org.scalajs.linker.runtime.RuntimeLong$.signedToDoubleApprox;I;I;D
[error] * org.scalajs.linker.runtime.RuntimeLong$.uintToDouble;I;D
[error] * scala.scalajs.js.timers.package$.setTimeout;D;Lscala.Function0;Lscala.scalajs.js.timers.SetTimeoutHandle
[error] * static scala.scalajs.js.timers.package$.scala$scalajs$js$timers$package$$$anonfun$setTimeout$1;Lscala.Function0;Ljava.lang.Object
[error] * scala.runtime.AbstractFunction0.apply$mcV$sp;V
[error] * scala.Function0.apply$mcV$sp;V
[error] * scala.concurrent.impl.Promise$DefaultPromise.tryFailure;Ljava.lang.Throwable;Z
[error] * munit.FunSuite.munitValueTransform;Lscala.Function0;Lscala.concurrent.Future
[error] * cats.effect.unsafe.SchedulerSuite.runtime;Lcats.effect.unsafe.IORuntime
[error] * cats.effect.RunnersPlatform.runtime;Lcats.effect.unsafe.IORuntime
[error] * scala.Product.$init$;V
[error] * scala.Product2.$init$;V
[error] * scala.Tuple2._1;Ljava.lang.Object
[error] * scala.Tuple2._2;Ljava.lang.Object
[error] * scala.concurrent.duration.package$.DurationLong;J;J
[error] * scala.concurrent.duration.package$DurationLong.nanos;Lscala.concurrent.duration.FiniteDuration
[error] * scala.concurrent.duration.DurationConversions.nanos;Lscala.concurrent.duration.FiniteDuration
[error] * scala.concurrent.duration.package$DurationLong.nanoseconds;Lscala.concurrent.duration.FiniteDuration
[error] * scala.concurrent.duration.DurationConversions.nanoseconds;Lscala.concurrent.duration.FiniteDuration
[error] * static java.util.concurrent.TimeUnit.NANOSECONDS;Ljava.util.concurrent.TimeUnit
[error] * java.util.concurrent.TimeUnit$.NANOSECONDS;Ljava.util.concurrent.TimeUnit
[error] * scala.concurrent.duration.package$DurationLong.durationIn;Ljava.util.concurrent.TimeUnit;Lscala.concurrent.duration.FiniteDuration
[error] * scala.concurrent.duration.package$DurationLong.scala$concurrent$duration$DurationLong$$n;J
[error] * scala.concurrent.duration.package$DurationLong$.durationIn$extension;J;Ljava.util.concurrent.TimeUnit;Lscala.concurrent.duration.FiniteDuration
[error] * scala.concurrent.duration.Duration$.apply;J;Ljava.util.concurrent.TimeUnit;Lscala.concurrent.duration.FiniteDuration
[error] * cats.effect.IO$.sleep;Lscala.concurrent.duration.FiniteDuration;Lcats.effect.IO
[error] * scala.concurrent.duration.package$.DurationInt;I;I
[error] * scala.concurrent.duration.package$DurationInt.millis;Lscala.concurrent.duration.FiniteDuration
[error] * scala.concurrent.duration.DurationConversions.millis;Lscala.concurrent.duration.FiniteDuration
[error] * scala.concurrent.duration.package$DurationInt.milliseconds;Lscala.concurrent.duration.FiniteDuration
[error] * scala.concurrent.duration.DurationConversions.milliseconds;Lscala.concurrent.duration.FiniteDuration
[error] * static java.util.concurrent.TimeUnit.MILLISECONDS;Ljava.util.concurrent.TimeUnit
[error] * java.util.concurrent.TimeUnit$.MILLISECONDS;Ljava.util.concurrent.TimeUnit
[error] * scala.concurrent.duration.package$DurationInt.durationIn;Ljava.util.concurrent.TimeUnit;Lscala.concurrent.duration.FiniteDuration
[error] * scala.concurrent.duration.package$DurationInt.scala$concurrent$duration$DurationInt$$n;I
[error] * scala.concurrent.duration.package$DurationInt$.durationIn$extension;I;Ljava.util.concurrent.TimeUnit;Lscala.concurrent.duration.FiniteDuration
[error] * static org.scalajs.linker.runtime.RuntimeLong.fromInt;I;J
[error] * org.scalajs.linker.runtime.RuntimeLong$.fromInt;I;J
[error] * org.scalajs.linker.runtime.RuntimeLong$.pack;I;I;J
[error] * cats.effect.IO.race;Lcats.effect.IO;Lcats.effect.IO
[error] * cats.effect.IO$.race;Lcats.effect.IO;Lcats.effect.IO;Lcats.effect.IO
[error] * cats.effect.IO$.asyncForIO;Lcats.effect.kernel.Async
[error] * scala.Function1.$init$;V
[error] * cats.effect.IO.map;Lscala.Function1;Lcats.effect.IO
[error] * scala.package$.Right;Lscala.util.Right$
[error] * scala.util.Right$.apply;Ljava.lang.Object;Lscala.util.Right
[error] * scala.Predef$.$conforms;Lscala.Predef$$less$colon$less
[error] * munit.Compare$.compareSupertypeWithSubtype;Lscala.Predef$$less$colon$less;Lmunit.Compare
[error] * munit.ComparePriority2.compareSupertypeWithSubtype;Lscala.Predef$$less$colon$less;Lmunit.Compare
[error] * munit.Compare$.defaultCompare;Lmunit.Compare
[error] * private munit.Compare$.anyEquality;Lmunit.Compare
[error] * munit.diff.DiffOptions$.default;Lmunit.diff.DiffOptions
[error] * munit.FunSuite.assertEquals;Ljava.lang.Object;Ljava.lang.Object;Lscala.Function0;Lmunit.Location;Lmunit.Compare;Lmunit.diff.DiffOptions;V
[error] * munit.Assertions.assertEquals;Ljava.lang.Object;Ljava.lang.Object;Lscala.Function0;Lmunit.Location;Lmunit.Compare;Lmunit.diff.DiffOptions;V
[error] * munit.FunSuite.failComparison$default$4;Lmunit.Clues
[error] * munit.Assertions.failComparison$default$4;Lmunit.Clues
[error] * munit.FunSuite.failComparison;Ljava.lang.String;Ljava.lang.Object;Ljava.lang.Object;Lmunit.Clues;Lmunit.Location;Lmunit.diff.DiffOptions;E
[error] * munit.FunSuite.assertEquals$default$3;Ljava.lang.Object
[error] * munit.Assertions.assertEquals$default$3;Ljava.lang.Object
[error] * cats.effect.IO$.realTime;Lcats.effect.IO
[error] * cats.effect.IO.flatMap;Lscala.Function1;Lcats.effect.IO
[error] * scala.collection.immutable.List$.canBuildFrom;Lscala.collection.generic.CanBuildFrom
[error] * scala.collection.generic.GenTraversableFactory.ReusableCBF;Lscala.collection.generic.GenTraversableFactory$GenericCanBuildFrom
[error] * scala.collection.immutable.List.map;Lscala.Function1;Lscala.collection.generic.CanBuildFrom;Ljava.lang.Object
[error] * scala.runtime.AbstractFunction1.$$Lambda$70e1780b84463d18653aacefee3ab989ac625f28.apply;Ljava.lang.Object;Ljava.lang.Object
[error] * scala.collection.immutable.$colon$colon.tl_$eq;Lscala.collection.immutable.List;V
[error] * scala.collection.TraversableLike.map;Lscala.Function1;Lscala.collection.generic.CanBuildFrom;Ljava.lang.Object
[error] * scala.collection.immutable.List.foreach;Lscala.Function1;V
[error] * munit.FunSuite.assert;Lscala.Function0;Lscala.Function0;Lmunit.Location;V
[error] * munit.Assertions.assert;Lscala.Function0;Lscala.Function0;Lmunit.Location;V
[error] * munit.FunSuite.munitCaptureClues;Lscala.Function0;Lscala.Tuple2
[error] * scala.Tuple2._1$mcZ$sp;Z
[error] * munit.FunSuite.munitPrint;Lscala.Function0;Ljava.lang.String
[error] * munit.Assertions.munitPrint;Lscala.Function0;Ljava.lang.String
[error] * munit.FunSuite.printer;Lmunit.diff.Printer
[error] * munit.Assertions.printer;Lmunit.diff.Printer
[error] * munit.FunSuite.fail;Ljava.lang.String;Lmunit.Clues;Lmunit.Location;Lmunit.diff.DiffOptions;E
[error] * scala.collection.immutable.List.exists;Lscala.Function1;Z
[error] * scala.collection.LinearSeqOptimized.exists;Lscala.Function1;Z
[error] * scala.concurrent.duration.FiniteDuration.toMicros;J
[error] * static org.scalajs.linker.runtime.RuntimeLong.and;I;I;I;I;J
[error] * org.scalajs.linker.runtime.RuntimeLong$.and;I;I;I;I;J
[error] * static org.scalajs.linker.runtime.RuntimeLong.sar;I;I;I;J
[error] * org.scalajs.linker.runtime.RuntimeLong$.sar;I;I;I;J
[error] * static org.scalajs.linker.runtime.RuntimeLong.mul;I;I;I;I;J
[error] * org.scalajs.linker.runtime.RuntimeLong$.mul;I;I;I;I;J
[error] * static org.scalajs.linker.runtime.RuntimeLong.shr;I;I;I;J
[error] * org.scalajs.linker.runtime.RuntimeLong$.shr;I;I;I;J
[error] * static org.scalajs.linker.runtime.RuntimeLong.add;I;I;I;I;J
[error] * org.scalajs.linker.runtime.RuntimeLong$.add;I;I;I;I;J
[error] * static org.scalajs.linker.runtime.RuntimeLong.sub;I;I;I;I;J
[error] * org.scalajs.linker.runtime.RuntimeLong$.sub;I;I;I;I;J
[error] * static org.scalajs.linker.runtime.RuntimeLong.notEquals;I;I;I;I;Z
[error] * org.scalajs.linker.runtime.RuntimeLong$.notEquals;I;I;I;I;Z
[error] * org.scalajs.linker.runtime.RuntimeLong$.equals;I;I;I;I;Z
[error] * munit.FunSuite.assert$default$2;Ljava.lang.Object
[error] * munit.Assertions.assert$default$2;Ljava.lang.Object
[error] * cats.effect.IO$.apply;Lscala.Function0;Lcats.effect.IO
[error] * cats.effect.IO$.delay;Lscala.Function0;Lcats.effect.IO
[error] * cats.effect.tracing.TracingPlatform.calculateTracingEvent;Lscala.Function0;Lcats.effect.tracing.TracingEvent
[error] * static java.lang.System.currentTimeMillis;J
[error] * java.lang.System$.currentTimeMillis;J
[error] * static org.scalajs.linker.runtime.RuntimeLong.fromDouble;D;J
[error] * static org.scalajs.linker.runtime.RuntimeLong.le;I;I;I;I;Z
[error] * org.scalajs.linker.runtime.RuntimeLong$.le;I;I;I;I;Z
[error] * org.scalajs.linker.runtime.RuntimeLong$.inlineUnsignedInt_$less$eq;I;I;Z
[error] * cats.effect.unsafe.IORuntime.scheduler;Lcats.effect.unsafe.Scheduler
[error] * java.util.concurrent.atomic.AtomicBoolean.set;Z;V
```

</details>

To reproduce, checkout this PR (https://github.com/typelevel/cats-effect/pull/4542) and run `sbt +testsJS/test`.

--- a/linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala
+++ b/linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala
@@ -834,7 +834,7 @@ private[optimizer] abstract class OptimizerCore(
       .withLocalDefs(paramLocalDefs)
       .withLocalDefs(restParamLocalDef.toList)
 
-    transformCapturingBody(captureParams, tcaptureValues, body, innerEnv) {
+    transformCapturingBody(captureParams, tcaptureValues, resultType, body, innerEnv) {
       (newCaptureParams, newCaptureValues, newBody) =>
         val newClosure = {
           Closure(flags, newCaptureParams, newParams, newRestParam, resultType,
@@ -845,7 +845,7 @@ private[optimizer] abstract class OptimizerCore(
   }
 
   private def transformCapturingBody(captureParams: List[ParamDef],
-      tcaptureValues: List[PreTransform], body: Tree, innerEnv: OptEnv)(
+      tcaptureValues: List[PreTransform], resultType: Type, body: Tree, innerEnv: OptEnv)(
       inner: (List[ParamDef], List[Tree], Tree) => PreTransTree)(cont: PreTransCont)(
       implicit scope: Scope, pos: Position): TailRec[Tree] = {
     /* Process captures.
@@ -913,7 +913,7 @@ private[optimizer] abstract class OptimizerCore(
 
     val innerScope = scope.withEnv(innerEnv.withLocalDefs(captureParamLocalDefs.result()))
 
-    val newBody = transformExpr(body)(innerScope)
+    val newBody = transform(body, isStat = resultType == VoidType)(innerScope)
 
     withNewLocalDefs(captureValueBindings.result()) { (localDefs, cont1) =>
       val (finalCaptureParams, finalCaptureValues) = (for {
@@ -2540,7 +2540,8 @@ private[optimizer] abstract class OptimizerCore(
 
             val methodDef = getMethodBody(targetMethod)
 
-            transformCapturingBody(methodDef.args, targs, methodDef.body.get, OptEnv.Empty) {
+            transformCapturingBody(methodDef.args, targs, AnyType,
+                methodDef.body.get, OptEnv.Empty) {
               (newCaptureParams, newCaptureValues, newBody) =>
                 if (!importReplacement.used.value.isUsed)
                   cancelFun()

--- a/test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/RegressionTest.scala
+++ b/test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/RegressionTest.scala
@@ -982,6 +982,19 @@ class RegressionTest {
     assertEquals("g", hide(b.bar))
   }
 
+  @Test
+  def typedClosureWithStatementBodyMustBeTransformedAsStat_Issue5331(): Unit = {
+    final class Box(var x: Int)
+
+    @noinline
+    def escapeAndCall(body: Runnable): Unit = body.run()
+
+    val box = new Box(5)
+    val task: Runnable = () => box.x = 6 // SAM for Runnable
+    escapeAndCall(task)
+    assertEquals(6, box.x)
+  }
+
 }
 
 object RegressionTest {

